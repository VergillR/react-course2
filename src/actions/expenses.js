// import uuid from 'uuid'
import db from '../firebase/firebase'

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
})

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '', note = '', amount = 0, createdAt = 0
    } = expenseData
    const newExpense = { description, note, amount, createdAt }

    return db.ref('expenses').push(newExpense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...newExpense
      }))
    })
  }
}

// REMOVE_EXPENSE
export const removeExpense = ({ id = '' } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
})

// EDIT_EXPENSE
export const editExpense = (id = '', updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
})

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
})

// export const startSetExpenses
// 1. fetch all expense data now
// 2. parse that data into an array
// 3. dispatch SET_EXPENSES
export const startSetExpenses = () => {
  return (dispatch) => {
    return db.ref('expenses').once('value', (snapshot) => {
      const temp = []
      snapshot.forEach((childSnapshot) => {
        temp.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        })
      })
      dispatch(setExpenses(temp))
    })
  }
}

export const startRemoveExpense = (exp = {}) => {
  return (dispatch) => {
    return db.ref(`expenses/${exp.id}`).remove().then(() => {
      dispatch(removeExpense(exp))
    }).catch((err) => {
      console.log(err)
    })
  }
}
