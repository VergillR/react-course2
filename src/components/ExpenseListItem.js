// Export a stateless functional component
// description, amount, createdAt
import React from 'react'
import { Link } from 'react-router-dom'
import moment from 'moment'
import numeral from 'numeral'

// load a locale
numeral.register('locale', 'nl', {
  delimiters: {
    thousands: '.',
    decimal: ','
  },
  abbreviations: {
    thousand: 'k',
    million: 'mln',
    billion: 'mrd',
    trillion: 'bln'
  },
  ordinal: function (number) {
    var remainder = number % 100
    return (number !== 0 && remainder <= 1 || remainder === 8 || remainder >= 20) ? 'ste' : 'de'
  },
  currency: {
    symbol: 'â‚¬ '
  }
})

// switch between locales
numeral.locale('nl')

const ExpenseListItem = ({ description, amount, createdAt, id }) => (
  <div>
    <h2>Expense</h2>
    <Link to={'/edit/' + id}><h3>Description: {description}</h3></Link>
    <p>Amount: {numeral(amount / 100).format('$0,0.00')}</p>
    <p>CreatedAt: {moment(createdAt).format('MMMM Do, YYYY')}</p>
  </div>
)

export default ExpenseListItem
