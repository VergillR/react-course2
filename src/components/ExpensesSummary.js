import React from 'react'
import { connect } from 'react-redux'
import getTotalExpenses from '../selectors/expenses-total'
import visibleExpenses from '../selectors/expenses'
import numeral from 'numeral'

// load a locale
numeral.register('locale', 'nl-nl', {
  delimiters: {
    thousands: '.',
    decimal: ','
  },
  abbreviations: {
    thousand: 'k',
    million: 'mln',
    billion: 'mrd',
    trillion: 'bln'
  },
  ordinal: function (number) {
    var remainder = number % 100
    return (number !== 0 && remainder <= 1 || remainder === 8 || remainder >= 20) ? 'ste' : 'de'
  },
  currency: {
    symbol: 'â‚¬ '
  }
})

// switch between locales
numeral.locale('nl-nl')

export class ExpensesSummary extends React.Component {
  /*
  Viewing { visibleExpenses(this.props.expenses, this.props.filters).length } expense(s) totalling { numeral(getTotalExpenses(visibleExpenses(this.props.expenses, this.props.filters)) / 100).format('$0,0.00') }

*/

  render () {
    return (
      <div>
        <p>
          Viewing { this.props.selection.length } { this.props.selection.length === 1 ? 'expense' : 'expenses' } totalling { numeral(getTotalExpenses(this.props.selection) / 100).format('$0,0.00') }
        </p>
      </div>
    )
  }
}

const mapStateToProps = (state, props) => ({
  selection: visibleExpenses(state.expenses, state.filters)
})

const mapDispatchToProps = (dispatch, props) => ({

})

export default connect(mapStateToProps, mapDispatchToProps)(ExpensesSummary)
