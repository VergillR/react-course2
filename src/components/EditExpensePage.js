/* eslint-disable react/prop-types */
import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { editExpense, removeExpense } from '../actions/expenses'

export class EditExpensePage extends React.Component {
  constructor (props) {
    super(props)
    this.onSubmit = (expense) => {
      props.editExpense(props.expense.id, expense)
      props.history.push('/')
    }
    this.onClick = (e) => {
      props.removeExpense({id: props.expense.id})
      props.history.push('/')
    }
  }

  render () {
    return (
      <div>
        <h3>Edit Expense</h3>
        <ExpenseForm expense={this.props.expense} onSubmit={(this.onSubmit)} />
        <button onClick={this.onClick}>Remove</button>
      </div>
    )
  }
}

const mapStateToProps = (state, props) => {
  // use props of the component itself to check what it needs from Redux store
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  }
}

const mapDispatchToProps = (dispatch, props) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: (data) => dispatch(removeExpense(data))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)
