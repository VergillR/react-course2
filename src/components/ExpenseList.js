import React from 'react'
import { connect } from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'
import selectTotalExpenses from '../selectors/expenses-total'
import numeral from 'numeral'

// switch between locales
numeral.locale('nl')

export const ExpenseList = (props) => (
  <div>
    { props.expenses.length === 0 ? <p>No expenses found</p> : props.expenses.map((expense) => <ExpenseListItem {...expense} key={expense.id} />) }
  </div>
)

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters),
    totalExpenses: selectTotalExpenses(state.expenses)
  }
}

export default connect(mapStateToProps)(ExpenseList)
