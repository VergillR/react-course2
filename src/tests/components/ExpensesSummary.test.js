/* global test, expect */
import React from 'react'
import { shallow } from 'enzyme'
import expenses from '../fixtures/expenses'
import visibleExpenses from '../../selectors/expenses'
import { ExpensesSummary } from '../../components/ExpensesSummary'
import moment from 'moment'

const emptyFilters = {
  text: '',
  sortBy: 'date',
  startDate: undefined,
  endDate: undefined
}

const altFilters = {
  text: 't',
  sortBy: 'amount',
  startDate: moment(0),
  endDate: moment(0).add(3, 'days')
}

test('should show all expenses with empty filters', () => {
  const wrapper = shallow(<ExpensesSummary selection={visibleExpenses(expenses, emptyFilters)} />)
  expect(wrapper).toMatchSnapshot()
})

test('should only show valid expenses with alt filters', () => {
  const wrapper = shallow(<ExpensesSummary selection={visibleExpenses(expenses, altFilters)} />)
  expect(wrapper).toMatchSnapshot()
})
