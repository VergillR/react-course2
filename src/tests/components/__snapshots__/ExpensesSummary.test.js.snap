// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should only show valid expenses with alt filters 1`] = `
<div>
  <p>
    Viewing 
    1
     
    expense
     totalling 
    € 45,00
  </p>
</div>
`;

exports[`should show all expenses with empty filters 1`] = `
<div>
  <p>
    Viewing 
    3
     
    expenses
     totalling 
    € 115,00
  </p>
</div>
`;
